### PR TITLE
Resource Naming DependentUpon Convention

### LINKED ISSUE #4488: Support resource naming based on DependentUpon source by convention
Related issues:

https://github.com/dotnet/winforms/issues/638
https://github.com/dotnet/project-system/issues/4807
https://github.com/dotnet/corefx/issues/34190

In SDK projects and new project system, we have eliminated the need for DependentUpon to make the appropriate file nesting in the IDE tree. However, there is a place where the build actually uses DependentUpon to:

1. Locate a source file, and _parse (!)_ it to get first class name and namespace
2. Generate .resources accordingly.

Let first me say that this whole design is terrible and it is absolutely insane that we have code to find the class and namespace name in msbuild:

https://github.com/microsoft/msbuild/blob/f309171ed821eb66cb635403e1fbff41ea22ae9c/src/Tasks/CSharpParserUtilities.cs#L15-L20

But people and features in VS rely on this and it has been a consistent source of feedback in moving to .NET Core 3.0.  So I give up and propose:

1. Add a boolean property that, when true, causes us to look for the source file by convention when DependentUpon metadata is not found. 

2. Set that property by default for .NET Core App 3.0 TFM.

cc @rainersigwald @davkean @livarcocc

### PR DESCRIPTION
**Pull request by BenVillalobos**
### Description

In SDK projects and new project system, we have eliminated the need for DependentUpon to make the appropriate file nesting in the IDE tree. However, there is a place where the build actually uses DependentUpon to:

1. Locate a source file, and _parse (!)_ it to get first class name and namespace
2. Generate .resources accordingly.

But people and features in VS rely on this and it has been a consistent source of feedback in moving to .NET Core 3.0. This opts into new behavior to use a convention instead of requiring explicit metadata.

Original issue: https://github.com/microsoft/msbuild/issues/4488

 This change is dependent upon: https://github.com/dotnet/sdk/pull/3533

### Customer Impact

In a fairly common situation for projects that use resources, avoids the need to specify `DependentUpon` metadata for each resource.

### Regression?

No

### Risk

Low risk.

### Test changes in this PR

------------------------------

Fixes #4488 

Added new argument to CreateManifestResourceName: `UseDependentUponConvention`

### CODE REVIEW

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs [hunk 1]
```diff
@@ -377,6 +386,138 @@ public void Regress188319()
             Assert.Equal(@"CustomToolTest.SR1", resourceNames[0].ItemSpec);
         }
 
+        /// <summary>
+        /// Opt into DependentUpon convention and load the expected file properly.
+        /// </summary>
+        [Fact]
+        public void DependentUpon_UseConvention()
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]: Maybe 
```suggestion
        public void DependentUponConvention_FindsMatch()
```

?

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs [hunk 2]
```diff
@@ -377,6 +386,138 @@ public void Regress188319()
...
+                CreateCSharpManifestResourceName t = new CreateCSharpManifestResourceName
+                {
+                    BuildEngine = new MockEngine(_testOutput),
+                    UseDependentUponConvention = true,
+                    ResourceFiles = new ITaskItem[] { i }
+                };
+
+                t.Execute().ShouldBeTrue("Expected the task to succeed.");
+
+                t.ManifestResourceNames.Length.ShouldBe(1, "Only 1 resource file should exist");
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]:
```suggestion
                t.ManifestResourceNames.ShouldHaveSingleItem();
```

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs [hunk 3]
```diff
@@ -377,6 +386,138 @@ public void Regress188319()
...
+            }
+        }
+
+
+
+        /// <summary>
+        /// When disabling UseDependentUponConvention it will find no .cs file and default to filename.
+        /// </summary>
+        [Fact]
+        public void DependentUpon_DisableConvention()
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]:
```suggestion
        public void DependentUponConvention_ConventionDisabledDoesNotReadConventionFile()
```

? It's a mouthful!

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs (hunk 2 above)
Comment (rainersigwald) [reply src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]: (for all of them)

**Comment by nguerrera** [comment]
cc @davkean

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs [hunk 4]
```diff
@@ -377,6 +386,138 @@ public void Regress188319()
...
+                t.Execute().ShouldBeTrue("Expected the task to succeed.");
+
+                t.ManifestResourceNames.ShouldHaveSingleItem();
+
+                t.ManifestResourceNames[0].ItemSpec.ShouldBe("MyStuff2.Namespace.Class2", "Expected the namespace & class of SR2.");
+            }
+        }
+
+
+
```
Comment (nguerrera) [inline src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]: nit: Extra newlines here.

File: src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs [hunk 5]
```diff
@@ -377,6 +386,138 @@ public void Regress188319()
...
+                t.Execute().ShouldBeTrue("Expected the task to succeed.");
+
+                t.ManifestResourceNames.ShouldHaveSingleItem();
+
+                t.ManifestResourceNames[0].ItemSpec.ShouldBe("SR1", "Expected only the file name.");
+            }
+        }
+
+
```
Comment (nguerrera) [inline src/Tasks.UnitTests/CreateCSharpManifestResourceName_Tests.cs]: nit: extra newline here.

File: src/Tasks/CreateManifestResourceName.cs [hunk 6]
```diff
@@ -142,6 +146,17 @@ CreateFileStream createFileStream
                     string fileName = resourceFile.ItemSpec;
                     string dependentUpon = resourceFile.GetMetadata(ItemMetadataNames.dependentUpon);
 
+                    // If opted into convention and no DependentUpon metadata, reference "<filename>.cs" if it exists.
+                    if(UseDependentUponConvention && string.IsNullOrEmpty(dependentUpon))
```
Comment (nguerrera) [inline src/Tasks/CreateManifestResourceName.cs]: nit
```suggestion
                    if (UseDependentUponConvention && string.IsNullOrEmpty(dependentUpon))
```

File: src/Tasks/CreateManifestResourceName.cs [hunk 7]
```diff
@@ -142,6 +146,17 @@ CreateFileStream createFileStream
...
+                    {
+                        string conventionDependentUpon = Path.ChangeExtension(fileName, SourceFileExtension);
+
+                        if(File.Exists(conventionDependentUpon))
```
Comment (nguerrera) [inline src/Tasks/CreateManifestResourceName.cs]: nit
```suggestion
                        if (File.Exists(conventionDependentUpon))
```

File: src/Tasks/CreateVisualBasicManifestResourceName.cs [hunk 8]
```diff
@@ -15,6 +15,8 @@ namespace Microsoft.Build.Tasks
     /// </summary>
     public class CreateVisualBasicManifestResourceName : CreateManifestResourceName
     {
+        public override string SourceFileExtension => ".vb";
```
Comment (nguerrera) [inline src/Tasks/CreateVisualBasicManifestResourceName.cs]: Is there an F# equivalent to this somewhere? cc @kevinransom

Comment (rainersigwald) [reply src/Tasks/CreateVisualBasicManifestResourceName.cs]: Looks like there is: https://github.com/dotnet/fsharp/blob/5f35d4e653d39bf5a5d8a18ac576f97eabe6c368/src/fsharp/FSharp.Build/CreateFSharpManifestResourceName.fs

**Comment by livarcocc** [comment]
Do we need any follow up validation from the WPF/WinForms team? Any value in having them trying out a private before merging?

**Comment by BenVillalobos** [comment]
@livarcocc it certainly couldn't hurt, but I wouldn't think it's necessary. @rainersigwald may think otherwise

**Comment by rainersigwald** [comment]
I think several folks are on vacation, but worth looping in @merriemcgaw @dreddy-work @zsd4yr.

**Comment by ericstj** [comment]
Did you test this with an item that has a partial path as it's ItemSpec?  /cc @anipik

**Comment by rainersigwald** [comment]
@ericstj are you hitting #4695? It should be fixed in the very latest builds (with #4702).

**Comment by ericstj** [comment]
Yep that appears to be it.  Thanks for the pointer.  For some reason my searching didn't find it.  I should have looked at referenced issues.  Thanks!

**Pull request merged**